### ISSUE TITLE
Content Script Organization for v1.0 (looking for feedback)

### ISSUE THREAD

**Issue by aklinker1** [wxt-dev/wxt#192]
This is the last breaking change I'm considering before releasing version 1.0.0.

Essentially, in a extension with lots of entrypoints, content scripts are mixed into the rest of the list. Instead of having them all inline with the `.content` suffix, it would probably be better to put them in a folder called `content-scripts`. This would make them easier to identify from the other entrypoints, and would mimic the output directory.

###### Current

```
entrypoints/content.[jt]sx?
entrypoints/content/index.[jt]sx?
entrypoints/<name>.content.[jt]sx?
entrypoints/<name>.content/index.[jt]sx?
```

###### Proposed

```
entrypoints/content.[jt]sx?
entrypoints/content-scripts/<name>.[jt]sx?
entrypoints/content-scripts/<name>/index.[jt]sx?
```

**Comment by dvlden** [comment]
I did not get too familiar with the docs _(just yet)_, but what you suggested sounds reasonable and clean.

I'm hoping that option to define `content.ts` in `entrypoints` without subdir `content-scripts` will remain possible?

Aside I do like the `content-scripts` as it is descriptive, but I hate `entrypoints`. Luckily, I think there's a config for that, so I can rename it for myself? Although, I'd suggest renaming it internally to just `entries`.

Best of luck! 🚀

**Comment by aklinker1** [comment]
@dvlden I can keep `content.ts`, it is very convenient for extensions with a single content script, thanks for the feedback!

> Luckily, I think there's a config for that, so I can rename it for myself? Although, I'd suggest renaming it internally to just entries.

Yes, [`entrypointsDir`](https://wxt.dev/api/wxt/interfaces/InlineConfig.html#entrypointsdir)

**Comment by molvqingtai** [comment]
I prefer to omit `entryopints`, like this:

```
src
src/assets
src/content-script
src/opitons
src/devtools
...
```
Or you can specify the paths of entrances such as `content-script`, `opitons`, `devtools`, etc.

**Comment by aklinker1** [comment]
Personally, I prefer flat project structures, with almost no subdirectories. I've found this type of project scales the best for web projects, and is adopted by many large frameworks (both Nuxt and Next).

```
src/
  assets/
  components/
  composables/
  entrypoints/
  public/
  utils/
```

The idea of the entrypoints directory is that it would mimic [Nuxt's `pages/` directory](https://nuxt.com/docs/guide/directory-structure/pages). This also makes auto-imports simple and easy to understand.

Obviously this is more relevant to projects using frontend frameworks, but I could see it working just fine with vanilla projects as well.

@molvqingtai I've never tried setting `entrypointsDir: "."`. Does it work out of the box with WXT?

```ts
export default defineConfig({
  srcDir: "src",
  entrypointsDir: ".",
});
```

You bring up a good point though, if I introduce a separate directory for content scripts, should I allow customization for the other directories? I think no, all entrypoints should go in a single folder to keep things consistent.

**Comment by molvqingtai** [comment]
> Personally, I prefer flat project structures, with almost no subdirectories other than the "main" ones. I've found this type of project scales the best for medium to large web projects, and is adopted by Nuxt (with some customization options).
> 
> ```
> src/
>   assets/
>   components/
>   composables/
>   entrypoints/
>   public/
>   utils/
> ```
> 
> This structure makes auto-imports simple and easy to understand - if it's in a main folders, it's available as an auto-import. The idea of the `entrypoints/` directory is to mimic [Nuxt's `pages/` directory](https://nuxt.com/docs/guide/directory-structure/pages).
> 
> If it's not obvious, this project is basically a clone of Nuxt but for Web Extensions lol. I use the same internal libraries, provide a similar DX, use same project layout, among other things. So I will not be supporting things that contradict a lot of the standards put forth by the Nuxt team.
> 
> @molvqingtai I've never tried setting `entrypointsDir: "."`. Does it work out of the box with WXT?
> 
> ```ts
> export default defineConfig({
>   srcDir: "src",
>   entrypointsDir: ".",
> });
> ```
> 
> Edit: No, it doesn't seem to work, at least not for the Vue template
> You bring up a good point though, if WXT introduces a separate directory for content scripts, should it allow customization for the other directories? I think no, all entrypoints should go in a single folder to keep things consistent.

Can you consider supporting `entrypointsDir: "."`

**Comment by aklinker1** [comment]
@molvqingtai I would be open to someone creating a PR to add support. The error message is very vague, and I have other priorities to work on right now.

**Comment by dvlden** [comment]
This should be closed, I assume?
#213 was tested by both of us and it works.

**Comment by aklinker1** [comment]
#213 doesn't address my original concern around putting content scripts in a separate folder, but I've decided to keep it as-is, with no separate folder for v1.0. Unless someone can change my mind, the code is much simpler and all entrypoints are in a single place, easy to find. It goes well with Nuxt's flat project structure.

**Comment by mefengl** [comment]
@aklinker1

Initially, I was going to ask why we aren't having a separate folder for content scripts, but the idea of a `flat project structure` really convinced me. I think it's indeed better.

Then I just realized, the reason I wanted a separate folder is that having multiple `name.content` files will clutter up my `entrypoints` folder.

![image](https://github.com/wxt-dev/wxt/assets/71683364/b11a26aa-ef06-4d92-a575-b3551dde117a)

So I am wondering why we are using `name.content` rather than `content.name`.

**Comment by mefengl** [comment]
@aklinker1

Initially, the absence of a separate folder for content scripts was a concern, but the concept of a `flat project structure` seems convincingly better.

The reason I wanted a separate folder is that having multiple `name.content` files will mess up my `entrypoints` folder.

![image](https://github.com/wxt-dev/wxt/assets/71683364/b11a26aa-ef06-4d92-a575-b3551dde117a)

Is there a specific reason to use `name.content` instead of `content.name`?

**Comment by aklinker1** [comment]
> @aklinker1
> 
> Initially, the absence of a separate folder for content scripts was a concern, but the concept of a `flat project structure` seems convincingly better.
> 
> The reason I wanted a separate folder is that having multiple `name.content` files will mess up my `entrypoints` folder.
> 
> ![image](https://github.com/wxt-dev/wxt/assets/71683364/b11a26aa-ef06-4d92-a575-b3551dde117a)
> 
> Is there a specific reason to use `name.content` instead of `content.name`?

This is exactly why I want to consider changing the naming convention. Content scripts not being placed next to each other can be annoying, and makes them hard to find.

Originally, I used suffixes because that seemed to be the standard when specifying the "type" of file. Think `*.test.ts`, `*.spec.ts`, `*.e2e.test.ts`, `*.module.css`, etc.

When it comes to organization, those suffixes work well because those files get placed next to the original file they're related to, which is the whole point.

For entrypoints, suffixes don't work because they don't put similar files next to one another, as shown in your screenshot. And this problem isn't just for content scripts, it's true for any type of entrypoint.

Prefixes are a solution, at the cost of a flat structure. Should the prefix just be `content-*` or `content.*`? In that case, files and folders will be grouped separately, and content scripts may still not be right next to each other.

So instead, if we supported a prefix of `content/`, a directory, it would put all them in the same place, but now we've lost the flat structure. maybe that's fine, maybe it isn't. Nuxt's `pages/` directory isn't flat, and that's the closest equivalent to WXT's `entrypoints/` directory.

---

TLDR: prefixes seem great, but they don't solve the organization problem completely unless we allow prefixed folders, like `content/`, which come at the cost of a non-flat directory structure.

**Comment by mefengl** [comment]
> Should the prefix just be content-* or content.*? In that case, files and folders will be grouped separately, and content scripts may still not be right next to each other.

Very important point, it makes me wonder if it's really worth it... In that case, it will look like this:

![](https://github.com/wxt-dev/wxt/assets/71683364/d81bb433-f59b-455b-9f9b-de061cd6a1e1)

But it also gives a chance to make things right, that is, use a folder for all entrypoints:

![](https://github.com/wxt-dev/wxt/assets/71683364/3bc9d6a6-0021-4bec-868b-a7b1016a1d19)

In fact, if only one content script is needed (which is enough for most extensions), then there will be no problem like this.

Adding to the previous discussion, if a prefix isn't the solution (I now think so too, as the *.content.name suffix indeed has its merits and can't address the folder and file organization issue), then there's no need to rush this decision. Since introducing a content folder (or content-script folder) wouldn't be a breaking change, it's not necessary to implement it before version 1.0.0.

**Comment by aklinker1** [comment]
Yeah, the main thing I want to avoid is providing too many ways to do something. Since this is an opinionated framework, there should be one correct way to accomplish one thing, so I don't want to keep adding additional patterns for naming entrypoints.

**Comment by aklinker1** [comment]
Didn't mean to close this.

**Comment by yunsii** [comment]
I don't know why place public dir in src dir? This is the first time I've seen it.

**Comment by aklinker1** [comment]
> I don't know why place public dir in src dir? This is the first time I've seen it.

@yunsii This project is based off Nuxt, which puts the `public/` directory inside the `srcDir` option: [nuxt-srcDir-example.zip](https://github.com/wxt-dev/wxt/files/13875886/nuxt-srcDir-example.zip)

> About entrypoints, I have a idea like Next.js [file-system based router](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts)...

Since the project is based off Nuxt, it makes more sense to consider [Nuxt's file based router](https://nuxt.com/docs/getting-started/routing). Basically the same, but no underscores. I looked through Next's docs, and couldn't figure out what the `_` meant in the file/folder names.

But that said, the `pages/` directory is not the same as `entrypoints/`. The pages directory is meant to mimic a URL path. Entrypoints are output in a flat structure in the root of the bundled extension. There are similarities, like they're both "entrypoints" for some view, but that's about it. This is why the current `entrypoints` structure doesn't allow nested directories, to better mimic the way files are output.

I do like the `*.entry.ts` idea if we did support nested entrypoints. That would make it easy to identify entrypoints vs other files associated with an entrypoint (other CSS or JS files).

Content scripts aren't special. There are other types of entrypoints, like sandbox, devtools, panels, and unlisted scripts/pages, that could grouped in a similar way when multiple are defined.

**Comment by aklinker1** [comment]
OK, after several peoples feedback, there are two asks:

1. Group entrypoints of the same type (content scripts, side panels, devtools, etc) next to each other in the file tree
2. Group related entrypoints together (ie: all code for a single domain)

In my mind, it's impossible to have both of these at the same time perfectly.

- If content scripts are all next to each other, it's impossible to place an unlisted page (like the UI for an iframe) or unlisted script (which might be injected into the page) next to them. 
- If related files (content script + iframe + injected scripts) are grouped together, then all the content scripts can't be grouped together

---

To support both of these to the best of my ability... I am not changing anything lol. The current setup provides each project some level of control over how entrypoints are organized.

You want to group content scripts together? Use a prefix:

```
entrypoints/
    overlay-google.content.ts
    overlay-duckduckgo.content.ts
    devtools.html
    devtools-panel.html
    devtools-pane.html
    iframe-overlay-google.html
    iframe-overlay-duckduckgo.html
```

Or you want to group all the related files together? Cool:

```
entrypoints/
    google-overlay.content.ts
    google-overlay-iframe.html
    duckduckgo-overlay.content.ts
    duckduckgo-overlay-iframe.html 
    devtools.html
    devtools-panel.html
    devtools-pane.html
```

Want a real world example?

<img width="300" alt="Screenshot 2024-01-09 at 9 56 06 AM" src="https://github.com/wxt-dev/wxt/assets/10101283/d6500ebd-59a4-4c78-ac2d-621afe5fe5fa">

https://github.com/anime-skip/player/tree/main/apps/extension/src/entrypoints

The one thing the current setup doesn't support is nested folders. I am deciding not to support them now.

If someone can propose a complete solution for organizing entrypoints, I'll consider it for v2 in the future. But the current setup will be used for v1.
